src/algebra/module/localized_module.lean

```diff
@@ -0,0 +1,358 @@
+  `mk m s + mk m' s' = mk (s' • m + s • m') (s * s')`
+* `localized_module.mk_smul_mk` : in the localized module, for any `r : R`, `s t : S`, `m : M`,
+  we have `mk r s • mk m t = mk (r • m) (s * t)` where `mk r s : localization S` is localized ring
+  by `S`.
+* `localized_module.is_module` : `localized_module M S` is a `localization S`-module.
+
+-/
+
+
+namespace localized_module
+
+universes u v
+
+variables {R : Type u} [comm_semiring R] (M : Type v) [add_comm_monoid M] [module R M]
+variables (S : submonoid R)
+
+example (r : R) (m : M) : M := r • m
+
+/--The equivalence relation on `M × S` where `(m1, s1) ≈ (m2, s2)` if and only if
+for some (u : S), u * (s2 • m1 - s1 • m2) = 0-/
+def r (p1 p2 : M × S) : Prop :=
+match p1, p2 with
+| ⟨m1, s1⟩, ⟨m2, s2⟩ := ∃ (u : S), u • s1 • m2 = u • s2 • m1
+end
+
+lemma r.is_equiv : is_equiv _ (r M S) :=
+{ refl := λ ⟨m, s⟩, ⟨1, by rw [one_smul]⟩,
+  trans := λ ⟨m1, s1⟩ ⟨m2, s2⟩ ⟨m3, s3⟩ ⟨u1, hu1⟩ ⟨u2, hu2⟩, begin
+    use u1 * u2 * s2,
+    rw calc (u1 * u2 * s2) • s1 • m3
+        = (u1 * u2 * s2 * s1) • m3 : by simp only [mul_smul]
+    ... = (u1 * s1 * (u2 * s2)) • m3 : by congr' 1; ext; simp only [submonoid.coe_mul]; ring
+    ... = (u1 * s1) • u2 • s2 • m3 : by simp only [mul_smul]
+    ... = (u1 * s1) • u2 • s3 • m2 : by rw [hu2]
+    ... = (u1 * s1 * (u2 * s3)) • m2 : by simp only [mul_smul]
+    ... = (u2 * s3 * (u1 * s1)) • m2 : by congr' 1; ext; simp only [submonoid.coe_mul]; ring
+    ... = (u2 * s3) • u1 • s1 • m2 : by simp only [mul_smul]
+    ... = (u2 * s3) • u1 • s2 • m1 : by rw [hu1]
+    ... = ((u2 * s3) * u1 * s2) • m1 : by simp only [mul_smul],
+    rw [← mul_smul],
+    congr' 1,
+    ext,
+    simp only [submonoid.coe_mul],
+    ring
+  end,
+  symm := λ ⟨m1, s1⟩ ⟨m2, s2⟩ ⟨u, hu⟩, ⟨u, hu.symm⟩ }
+
+
+instance r.setoid : setoid (M × S) :=
+{ r := r M S,
+  iseqv := ⟨(r.is_equiv M S).refl, (r.is_equiv M S).symm, (r.is_equiv M S).trans⟩ }
+
+/--
+If `S` is a multiplicative subset of a ring `R` and `M` an `R`-module, then
+we can localize `M` by `S`.
+-/
+@[nolint has_inhabited_instance]
+def _root_.localized_module : Type (max u v) := quotient (r.setoid M S)
+
+section
+variables {M S}
+
+/--The canonical map sending `(m, s) ↦ m/s`-/
+def mk (m : M) (s : S) : localized_module M S :=
+quotient.mk ⟨m, s⟩
+
+lemma mk_eq {m m' : M} {s s' : S} : mk m s = mk m' s' ↔ ∃ (u : S), u • s • m' = u • s' • m :=
+quotient.eq
+
+@[elab_as_eliminator]
+lemma induction_on {β : localized_module M S → Prop} (h : ∀ (m : M) (s : S), β (mk m s)) :
+  ∀ (x : localized_module M S), β x := λ x,
+begin
+  induction x using quotient.induction_on,
+  rcases x with ⟨m, s⟩,
+  exact h m s,
+end
+
+@[elab_as_eliminator]
+lemma induction_on₂ {β : localized_module M S → localized_module M S → Prop}
+  (h : ∀ (m m' : M) (s s' : S), β (mk m s) (mk m' s')) : ∀ x y, β x y := λ x y,
+begin
+  induction x using quotient.induction_on,
+  induction y using quotient.induction_on,
+  rcases ⟨x, y⟩ with ⟨⟨m, s⟩, ⟨m', s'⟩⟩,
```

Comment:
I thought this was only legal in `obtain`!

Reply:
I would be nice if there is a `rinduction` tactic to immediately destruct the variables